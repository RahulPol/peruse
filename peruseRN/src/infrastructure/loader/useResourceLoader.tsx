import {
  Roboto_100Thin,
  Roboto_300Light,
  Roboto_400Regular,
  Roboto_500Medium,
  Roboto_700Bold,
} from '@expo-google-fonts/roboto';
import * as Font from 'expo-font';
import * as SplashScreen from 'expo-splash-screen';
import { useCallback, useEffect, useState } from 'react';

export const useResourceLoader = () => {
  const [appIsReady, setAppIsReady] = useState(false);
  const [error, setError] = useState('');

  useEffect(() => {
    async function prepare() {
      try {
        await SplashScreen.preventAutoHideAsync();

        await Font.loadAsync({
          Roboto_100Thin,
          Roboto_300Light,
          Roboto_400Regular,
          Roboto_500Medium,
          Roboto_700Bold,
        });
      } catch (e) {
        // TODO: Add error handling for application errors
        if (e instanceof Error) {
          setError(e.message);
        } else {
          setError('Unexpected error: ' + e);
        }
      } finally {
        setAppIsReady(true);
      }
    }

    prepare();
  }, []);

  const onLayoutRootView = useCallback(async () => {
    if (appIsReady) {
      await SplashScreen.hideAsync();
    }
  }, [appIsReady]);

  return { appIsReady, error, onLayoutRootView };
};
