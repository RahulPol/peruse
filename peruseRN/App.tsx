import { StatusBar } from 'expo-status-bar';
import { ThemeProvider } from 'styled-components';

import StyledView from './src/components/utility/styled-view.component';
import { useResourceLoader } from './src/infrastructure/loader/useResourceLoader';
import { Navigation } from './src/infrastructure/navigation';
import { lightTheme as theme } from './src/infrastructure/theme';
import { AuthenticationContextProvider } from './src/services/authentication/authentication.context';

export default function App() {
  const { appIsReady, onLayoutRootView } = useResourceLoader();

  if (!appIsReady) {
    return null;
  }

  return (
    <>
      <ThemeProvider theme={theme}>
        <AuthenticationContextProvider>
          <StyledView onLayout={onLayoutRootView}>
            <Navigation />
          </StyledView>
        </AuthenticationContextProvider>

        <StatusBar style="auto" />
      </ThemeProvider>
    </>
  );
}
