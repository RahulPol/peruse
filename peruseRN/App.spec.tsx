import { cleanup, render } from '@testing-library/react-native';

import App from './App';
import { useResourceLoader } from './src/infrastructure/loader/useResourceLoader';

jest.doMock('./src/components/utility/styled-view.component', () => {
  const MockComponent = () => <div />;
  return MockComponent;
});

jest.mock('./src/infrastructure/navigation', () => ({
  Navigation: () => {
    const MockName = 'navigation mock';
    // @ts-ignore
    return <MockName testID="mock-navigation" />;
  },
}));

jest.mock('./src/infrastructure/loader/useResourceLoader');

afterEach(() => cleanup());

describe('<App>', () => {
  it('should render when fonts are loaded', () => {
    (useResourceLoader as jest.Mock).mockReturnValueOnce({
      appIsReady: true,
      errorInLoading: false,
      onLayoutRootView: jest.fn(),
    });

    const { toJSON } = render(<App />);

    expect(toJSON()).toBeDefined();
  });

  it('should not render when fonts are not loaded', () => {
    (useResourceLoader as jest.Mock).mockReturnValueOnce({
      appIsReady: false,
      errorInLoading: true,
      onLayoutRootView: jest.fn(),
    });
    const { toJSON } = render(<App />);

    expect(toJSON()).toEqual(null);
  });
});
