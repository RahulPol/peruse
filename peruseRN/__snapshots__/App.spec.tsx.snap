// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App> has one child 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Text>
    Welcome to Expo
  </Text>
</View>
`;
